### ISSUE TITLE
`cxx.export.coptions` neither passed to linking target nor to the pkg-config file

### ISSUE THREAD

**Issue by lyrahgames** [build2/build2#316]
Currently, I am creating a [build2 package](https://github.com/lyrahgames/build2-libigl) for the header-only library [libigl](https://github.com/libigl/libigl). This library contains at least one file whose inclusion results in the following error.
```
fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj
```
Based on this behavior and according to the upstream build steps, I added the flag `/bigobj` to the `cxx.export.coptions` of the library for MSVC-based builds to make sure any other library or executable using this library will use this flag to not fail compilation.
```
if ($cxx.target.system == 'win32-msvc')
  lib{igl-core}: cxx.export.coptions += /bigobj
```
Unfortunately, the content of `cxx.export.coptions` neither was passed to the test executable nor to the pkg-config file. Could this be a bug? I also tried with unconditional assignments and other options, like `lib{igl-core}: cxx.export.coptions = -O2`, which also did not work.

**Comment by Klaim** [comment]
I managed to confirm the issue with that project.

My understanding is that `build2` will not export `*.coptions` on purpose (so it would not be a bug, but maybe a feature request).

Therefore an easy fix is to pass the flag to the exported pre-processor values: https://github.com/lyrahgames/build2-libigl/pull/2/files#diff-14a2c21606b0e5c9bd268b1e3dfcedca4506da6df38859a10282b12de71413daL13-L18

These changes works for me on Windows/msvc.

**Comment by Klaim** [comment]
As discussed [there](https://github.com/lyrahgames/build2-libigl/pull/2) the solution to pass the flag to pre-processor values doesnt work because there is no binary and `build2` will not use the `Cflags: ...` from header-only libraries pkg-config files, so in the installed form the flags to pass to users are lost. Or at least so far I didnt manage to see anything I put in that field appear in the build command for an executable depending on it.

**Comment by boris-kolpackov** [comment]
This is by design (there is no `*.export.coptions`), see: https://github.com/build2/HOWTO/blob/master/entries/compile-options-in-buildfile.md, especially a note at the bottom.

As to what to do in this case, passing `/bigobj` in `*.export.poptions` is definitely wrong (per the rationale provided in the above howto). In particular, this would mean that every translation unit of a binary (executable or library) that depends on this library would be compiled with this option, whether it actually includes this library or not.

Unfortunately, there is no good "right" solution to this problem, except changing the library (preferably by upstream) not to require this option. Other than that, the options are:

1. Tell the user in the documentation that if they want to use this library with MSVC, then they will have to compile the relevant translation units with this option.

2. Still export it via `*.export.poptions`, even if this is wrong/suboptimal.

**Comment by lyrahgames** [comment]
> This is by design (there is no `*.export.coptions`), see: https://github.com/build2/HOWTO/blob/master/entries/compile-options-in-buildfile.md, especially a note at the bottom.

I see. Thanks for the clarification. I have not been aware of this fact.

> As to what to do in this case, passing `/bigobj` in `*.export.poptions` is definitely wrong (per the rationale provided in the above howto). In particular, this would mean that every translation unit of a binary (executable or library) that depends on this library would be compiled with this option, whether it actually includes this library or not.

I agree. It would be totally suboptimal to require this for all sources. But I also think that it is suboptimal to require the user to manually add those options by reading the documentation.
On the other hand, using `*.export.poptions` as a workaround feels to me like that we should have used `*.export.coptions` anyway to not conceal the actual intention.
Especially, I think of header-only libraries. It seems there is no proper way to provide any compile options for these to the user, is there? The only way I can think of is use the new `buildfile` export/import feature to export predefined compile option variables for certain files and  to access these variables in the importing project by also importing the `buildfile`.
Probably, it is completely infeasible to use file-based exportation.
Do you have any other ideas?

**Comment by boris-kolpackov** [comment]
> Especially, I think of header-only libraries. It seems there is no proper way to provide any compile options for these to the user, is there?

No, there isn't. But the point being is that a library should not impose any compile options on the consumer. Doing that opens all kinds of cans of worms, for example, what if two libraries impose contradictory options, what should happen in this case?

> The only way I can think of is use the new `buildfile` export/import feature to export predefined compile option variables for certain files and to access these variables in the importing project by also importing the `buildfile`.

Yes, this could be a some sort of "setup the environment for consuming the library" mechanism for libraries that cannot be made to work otherwise. At least it will be explicit and so the user will be aware of the potential issues.
